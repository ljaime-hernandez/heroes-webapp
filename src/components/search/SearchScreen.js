import queryString from 'query-string';
import { useMemo } from 'react';
import {useNavigate, useLocation} from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { getHeroesByName } from '../../selectors/getHeroesByName';
import { HeroCard } from '../hero/HeroCard';

export const SearchScreen = () => {

    // we use the useNavigate method from the react-router-dom library to assign its content on a 
    // constant which we will later use for a submit event
    const navigate = useNavigate();
    // the useLocation method will return us the URL which we are using at that very moment, helpful
    // for situations where we need to take specific information from it and retrieve arguments
    const location = useLocation();
    // the queryString library is used to separate all the arguments used on a query in a URL,
    // along with the useLocation method we can separate on strings the words or strings used on it,
    // in this case to use it as argument for a selector
    const {q = ''} = queryString.parse(location.search);
    // we use our custom hook to assign the query string value retrieved with the useLocation and
    // queryString to assign it on the form value, the handleInputChange function is used on the
    // change event to handle any modifications in the input text
    const [{searchText}, handleInputChange, reset] =  useForm({
        searchText: q
    })
    // the useMemo will allow us to keep information from the return of the getHeroesByName selector,
    // this will be highly helpful in situations where the function returns thousands of information 
    // pieces. The HeroBySearch variable will retain all the values returned by the function, this is meant
    // so in case we make another function call, if pieces of information are still the same as in the 
    // previous function call, the elements rendered will only be the new ones, adjusting to the old ones
    // but the old ones are not rendered again, saving useful memory process
    const heroesBySearch = useMemo(() => getHeroesByName(q), [q]);

    const handleSearch = (e) => {
        e.preventDefault();

        // the navigate will concatenate a query string with the actual string used to search for an specific
        // superhero in our data files, this will trigger a series of actions described from the top of this file to the bottom
        navigate(`?q=${searchText}`);
    }


    return (
        <>
            <h1>Search</h1>
            <hr/>

            <div className='row'>
                <div className='col-5'>
                    <h4>Type hero info</h4>

                    <form onSubmit={handleSearch}>
                        <input
                            type='text'
                            placeholder='Insert information here'
                            className='form-control'
                            name='searchText'
                            autoComplete='off'
                            value={searchText}
                            onChange={handleInputChange}
                        />

                        <button
                            className='btn btn-outline-primary mt-2 col-12'
                            type='submit'
                        >
                            Search
                        </button>

                    </form>
                </div>
                <div className='col-7'>
                    <h4>Results</h4>
                    <hr/> 

                    {
                        // if the query string is empty, an alert box will display a message prompting us
                        // to search, if there is a string but the string does not match any of the
                        // heroes in our data files, then the string will be rendered with another alert
                        // box clearly stating theres no match with the input 
                        (q === '')
                            ? <div className='alert alert-info'>Search for hero</div>
                            : (heroesBySearch.length === 0)
                                && <div className='alert alert-danger'>No results for: {q}</div>
                    }

                    {
                        // if theres any object contained in the heroesBySearch variable then its
                        // information will be rendered by the HeroCard component, one by one using
                        // the map method. If theres no objects in the heroesBySearch variable then
                        // the page wont render any hero but will render the information contained in the
                        // condition on top of this scope
                        heroesBySearch.map( hero => (
                            <HeroCard
                                key={hero.id}
                                {...hero}    
                            />
                        ))
                    }    
                </div>
            </div>
        </>
    )
}
