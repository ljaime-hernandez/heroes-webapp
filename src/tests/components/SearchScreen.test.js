import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { SearchScreen } from '../../components/search/SearchScreen'; 

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({

    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate
}));

describe('Tests on SearchScreen component', () => {

    test('should render properly with default values', () => {
        
        const wrapper = mount (
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen/>
            </MemoryRouter>)

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Search for hero');
    })

    test('should render properly with query parameter', () => {
        
        const wrapper = mount (
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchScreen/>
            </MemoryRouter>)

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('input').prop('value')).toBe('batman');
    })
    
    test('should render alert-danger class when the query does not match any hero', () => {
        
        const q = 'pokemon';

        const wrapper = mount (
            <MemoryRouter initialEntries={[`/search?q=${q}`]}>
                <SearchScreen/>
            </MemoryRouter>)

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-danger').text().trim()).toBe(`No results for: ${q}`);
    })

    test('should call navigate function to render new path', () => {
        
        const wrapper = mount (
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen/>
            </MemoryRouter>);

        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: 'superman'
            }});

            wrapper.find('form').prop('onSubmit')({
                preventDefault(){}
            })

            expect(mockNavigate).toHaveBeenCalledWith('?q=superman');
    })
    
    
})