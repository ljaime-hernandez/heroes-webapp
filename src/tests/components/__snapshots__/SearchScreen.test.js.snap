// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on SearchScreen component should render alert-danger class when the query does not match any hero 1`] = `
Array [
  <h1>
    Search
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Type hero info
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Insert information here"
          type="text"
          value="pokemon"
        />
        <button
          className="btn btn-outline-primary mt-2 col-12"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        No results for: 
        pokemon
      </div>
    </div>
  </div>,
]
`;

exports[`Tests on SearchScreen component should render properly with default values 1`] = `
Array [
  <h1>
    Search
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Type hero info
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Insert information here"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-2 col-12"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search for hero
      </div>
    </div>
  </div>,
]
`;

exports[`Tests on SearchScreen component should render properly with query parameter 1`] = `
Array [
  <h1>
    Search
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Type hero info
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Insert information here"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-primary mt-2 col-12"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="superhero"
                className="card-img"
                src="/assets/dc-batman.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  More...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
